cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java

```diff
@@ -0,0 +1,149 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closeables;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.Flushable;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Manage locations for {@link RollingLocationLogAppender}
+ */
+public class LocationManager implements Flushable, Closeable {
+  protected static final String TAG_NAMESPACE_ID = ".namespaceId";
+  protected static final String TAG_APPLICATION_ID = ".applicationId";
+
+  private Location logBaseDir;
+  private Map<LocationIdentifier, LocationOutputStream> activeLocations;
+  private String filePermissions;
+
+  public LocationManager(LocationFactory locationFactory, String basePath, String filePermissions)
+    throws IOException {
+    this.logBaseDir = locationFactory.create(basePath);
+    this.activeLocations = new HashMap<>();
+    this.filePermissions = filePermissions;
+  }
+
+  LocationIdentifier getLocationIdentifier(Map<String, String> propertyMap) throws IllegalArgumentException,
+    IOException {
+
+    String namespaceId = propertyMap.get(TAG_NAMESPACE_ID);
+    Preconditions.checkArgument(propertyMap.containsKey(TAG_APPLICATION_ID),
+                                String.format("%s is expected but not found in the context %s",
+                                              TAG_APPLICATION_ID, propertyMap));
+    String application = propertyMap.get(TAG_APPLICATION_ID);
+
+    return new LocationIdentifier(namespaceId, application);
+  }
+
+  OutputStream getLocationOutputStream(final LocationIdentifier locationIdentifier, String fileName)
+    throws IOException {
+    if (activeLocations.containsKey(locationIdentifier)) {
+      return activeLocations.get(locationIdentifier).getOutputStream();
+    }
+
+    Location logFile = getLogLocation(locationIdentifier).append(fileName);
+    Location logDir = Locations.getParent(logFile);
+    // check if parent directories exist
+    Locations.mkdirsIfNotExists(logDir);
+
+    if (logFile.exists()) {
+      // The file name for a given application exists if the appender was stopped and then started again but file was
+      // not rolled over. In this case, since the roll over size is typically small, we can rename the old file and
+      // copy its contents to new file and delete old file
+      long now = System.currentTimeMillis();
+      // rename existing file to temp file
+      if (logDir == null) {
+        // this should never happen
+        throw new IOException(String.format("Parent Directory for %s is null", logFile.toURI().toString()));
+      }
+      Location tempLocation = logFile.renameTo(logDir.append(Long.toString(now)));
```

Comment:
oh, this is just temp file, that you would copy to the active location output stream please add comments, and todo's for potential size limits.